[IMP] website, *: rename group publisher to restricted_editor

rdeodoo opened the pull request:
* portal, web_unsplash, website_*

This commit renames the `website.group_website_publisher` into
`website.group_website_restricted_editor`.

While the change in itself might look unuseful, it will help the dev and
tech community figuring which group is related to which feature.

Even internally when we discuss specs, we always have to remind which
group is the restricted editor: the publisher one or the designer one?

While it is probably, after all those years, now anchored in some dev
mind, there is no easy way to directly figure which of those 2 groups is
the restricted editor one.
Note that I myself always got confused about it.

Now, the "Restricted Editor" right will be reflected in its technical
name `group_website_restricted_editor`.
Same as for the "Editor & Designer" which technical name is
`group_website_designer`.

As we would like to have a fully working and ready system in v17 for the
community to be able to build themes easily, removing that dubious part
is a nice to have.

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/98200).

rdeodoo commented:
@qsm-odoo If you still don't object to this change, I'll mark it as ready and open the enterprise/upgrade PR.

qsm-odoo commented on addons/website/models/website.py:
```diff
@@ -1033,7 +1033,7 @@ def _force_website(self, website_id):
 
     @api.model
-    def is_publisher(self):
+    def is_restricted_editor(self):
         return self.env['ir.model.access'].check('ir.ui.view', 'write', False)
```
Was already the case before but strange that we have a method `is_XXX` which can return `True` even if you don't have the `group_XXX` no? :grimacing:

rdeodoo replied:
Well, this method seems wrong actually.
It is not checking for restricted editor (or publisher before this PR and since its introduction with https://github.com/odoo/odoo/commit/1acf6d904fa8474b6ba59735ae57a0323465178d 9 years ago) but for full editor / designer.

Restricted editor / publisher don't have the write right on ir.ui.view, always been like that https://github.com/odoo/odoo/blob/1acf6d904fa8474b6ba59735ae57a0323465178d/addons/website/security/ir.model.access.csv

I guess there was already some confusion about the naming back then :shrug: 

I will have a look tomorrow, I guess we should rename it and make it check for the group instead of ir.ui.view write right :thinking:

qsm-odoo replied:
> Restricted editor / publisher don't have the write right on ir.ui.view, always been like that

Well, precisly but this is what the method is doing: check that you don't have the write access on ir.ui.view. So the method is probably used rightfully everywhere and we don't wan't to use has_group (although I guess we could, always been unclear when we want to do has_group or access check). But the name of the method is confusing: I guess that in all flows is_restricted_editor === has_group(restricted_editor) but technically I would expect that is_restricted_editor checks if you have the group as the name suggests.

rdeodoo replied:
(Considering the code before this PR, so let's not think about the new naming `is_restricted_editor()`)
What the method does, despite its `is_publisher()` name is actually indirectly checking if you are a `website_designer` and not a `website_publisher` :grimacing: (on top of that it also allow you do be considered as designer if you have write access on view from another module, which doesn't makes sens actually, I don't see why someone from module xxx not related to website but which gives the ir.ui.view write would suddenly be considered as designer (wrongly "publisher" according to the wrong/misleading method name) and would see stuff supposed to be seen only by the website builder admin).

At the end, this method should just be renamed in to `is_designer()`, or even better, simply removed and replaced by call to `has_group(designer)`, as I don't see any valid reason for someone with write right on ir.ui.view be considered as a website builder admin :grimacing:

rdeodoo replied:
Maybe we could also add a comment in stable version to mention that despite being named `is_publisher()`, it actually check if you are a `designer`, not a `publisher`.

qsm-odoo replied:
Ok I was reading the method wrong, hence the confusion. I somehow thought that `self.env['ir.model.access'].check('ir.ui.view', 'write', False)` was checking you *don't* have write access because of the misleading "False" x)

> At the end, this method should just be renamed in to is_designer(), or even better, simply removed 

Indeed, I would entirely remove it, it's useless

> and replaced by call to has_group(designer), as I don't see any valid reason [...]

I guess it has to be checked case by case, indeed someone with write access on views should not be considered a website builder admin but maybe that method was used in wrong cases not related to website builder, seems strange we would introduce that for no reason? Don't know :grimacing:

> Maybe we could also add a comment in stable version to mention that despite being named is_publisher(), it actually check if you are a designer, not a publisher.

Depending on what we do, I would add a deprecated comment indeed.

qsm-odoo commented on addons/website/data/digest_data.xml:
```diff
@@ -4,7 +4,7 @@
             <field name="name">Tip: Engage with visitors to convert them into leads</field>
             <field name="sequence">1500</field>
-            <field name="group_id" ref="website.group_website_publisher" />
+            <field name="group_id" ref="website.group_website_restricted_editor" />
```
Since you had to touch those lines, I would have removed the space before "/>" at the same time :smile: But I guess it's a matter of taste :)

qsm-odoo reviewed (commented):
I don't personally object but I think many will ^^ As you wish ;) But since you are renaming the "publisher" thing, won't you rename the "designer" thing?

rdeodoo commented:
> But since you are renaming the "publisher" thing, won't you rename the "designer" thing?

As discussed, I didn't plan to change it, for 2 reasons:
1. The technical name `website.group_website_designer` is in par with the regular name `Editor & Designer`.
2. Now that the "Restricted Editor" technical name will become `website.group_website_restricted_editor`, it will be easier to figure and should IMO be enough to avoid the confusion. Either we are talking about the restricted editor, or the other -> the designer.

rdeodoo commented:
@qsm-odoo I added a few small commits on the PR, I'll let you check.
The goal is just to make the code easier to read and figure.

I am wondering if the `is_restricted_editor()` and `is_designer()` methods shouldn't be on the `env`, as `is_superuser()`, `is_admin()` and `is_system()` already are. That would allow to also get rid of the `has_group(xx)` when we don't have `request` but only `self.env`.

qsm-odoo commented:
> That would allow to also get rid of the has_group(xx) when we don't have request but only self.env.

Not sure to understand. Can't we just use `has_group` everywhere?

rdeodoo commented:
We could, and I would be ok with that, we just need to decide.

I would intuitively got for `env.user.is_designer()` / `env.user.is_restricted_editor()` over `env.user.has_group('website.group_website_designer')` / `env.user.has_group('website.group_website_restricted_editor')` as I feel its:
- Easier to remember (kinda, less unsure if you remember correctly the group(s) name(s))
- Shorter
- Looks less technical
- Not really a convincing argument let alone but also: one single place to modify if we were to want to change the "is restricted editor" check behavior

All in all, as those group checks are quite often used, those utils method would make a bit of sense but either ways are fine to me TBH. I'll you pick one and I'll adapt the PR.

qsm-odoo commented:
@rdeodoo Fine by me, let's keep an util method then :+1: Just have to make sure it is what we want to use everywhere, like I said in the above conversation:

> I guess it has to be checked case by case, indeed someone with write access on views should not be considered a website builder admin but maybe that method was used in wrong cases not related to website builder, seems strange we would introduce that for no reason? Don't know :grimacing:

rdeodoo commented:
I checked and it makes sense for those few occurrences, I am just a bit unsure about the slides one.
I keep finding unused method/attributes here and there, I'll make a later pass on the whole code to try to clean that a bit.

About the util, would you move them to the `res.users` / `env` record?
Right now, the utils are only callable on a website record, which doesn't really makes sense as there is plenty of places where we can't and still do `has_group()`.
[IMHO](https://github.com/odoo/odoo/pull/98200#issuecomment-1220890093), by preference order:
- Move utils on the `env`, as `is_superuser()`, `is_admin()` and `is_system()` already are. That would allow to also get rid of the `has_group(xx)` when we don't have `request` but only `self.env`.
- Move utils on the `res.users`, same rationale and that's actually where the method from the env end up -> there is private method doing the has_group check there. Seems a bit overkill to keep the same pattern tho (public method on env calling private on res.users)
- (current state of the code here and master) Utils are on website and need to call `has_group` when no website available (model method etc)
- No util and call `has_group` everywhere

qsm-odoo commented:
@rdeodoo I let you decide, all 4 suggestions are ok to me. I am just afraid that with the first two solutions we would have people asking "why don't we add a method per group for all groups of Odoo while we are at it" ^^. Personally I would use `has_group` everywhere :grimacing:

rdeodoo commented:
Well, I really can't decide which one would be better, so I went for the less diff and removed the helper. Replaced it by the `has_group()` call as we do elsewhere when we don't have access to the website to call the helper.

While it is a bit more verbose and technical, it's true that we won't create a helper on `env` or `res.users` for every existing groups. I guess it is hard to draw a line.

This one should be ready.

qsm-odoo reviewed (approved):
@robodoo rebase-ff r+

robodoo commented:
Merge method set to rebase and fast-forward.

robodoo commented:
@rdeodoo @qsm-odoo linked pull request(s) odoo/enterprise#30691 not ready. Linked PRs are not staged until all of them are ready.

robodoo commented:
@rdeodoo @qsm-odoo linked pull request(s) odoo/upgrade#3808 not ready. Linked PRs are not staged until all of them are ready.

qsm-odoo commented:
@robodoo r+ please
